### PR TITLE
[quant][graphmode][fx][test] Add per op test for graph mode quant on fx

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43335 [quant][graphmode][fx] Add support for batchnorm relu
* #43334 [quant][graphmode][fx] Add support for batchnorm
* #43333 [quant][graphmode][fx] Add support for cat
* #43332 [quant][graphmode][fx] Add support for add relu
* #43331 [quant][graphmode][fx] Add support for add
* #43287 [quant][graphmode][fx] Add support for conv module + relu
* #43286 [quant] Make OP_LIST_TO_FUSER_METHOD public
* #43285 [quant][graphmode][fx] Add support for conv module
* **#43229 [quant][graphmode][fx][test] Add per op test for graph mode quant on fx**
* #43175 [quant][graphmode][fx] Add graph mode quantization on fx

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23201692](https://our.internmc.facebook.com/intern/diff/D23201692)

### CODE REVIEW

File: torch/testing/_internal/common_quantization.py [hunk 1]
```diff
@@ -358,6 +367,65 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
...
+            prepared = quantizer.prepare(graph, qconfig_dict)
+
+        prepared(*inputs)
+        qgraph = quantizer.convert(prepared)
+        qgraph_debug = quantizer.convert(prepared, debug=True)
+
+        result = qgraph(*inputs)
+        result_debug = qgraph_debug(*inputs)
+
+        assert (result - result_debug).abs().max() == 0, \
```
Comment (supriyar) [inline torch/testing/_internal/common_quantization.py:425]: use `self.assertEqual`?

File: torch/testing/_internal/common_quantization.py [hunk 2]
```diff
@@ -358,6 +367,65 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
...
+            ' and call module are supported right now'
+        modules = dict(graph_module.root.named_modules())
+        for node in graph_module.graph.nodes:
+            if node.op == 'call_function' and node.op == target_node[0] and node.target == target_node[1]:
+                return
+            elif node.op == 'call_module' and node.op == target_node[0] and type(modules[node.target]) == target_node[1]:
+                return
+
+        self.assertTrue(False, 'node:' + str(target_node) +
+                        ' not found in the graph module')
```
Comment (supriyar) [inline torch/testing/_internal/common_quantization.py:390]: nit: move to previous line?

File: torch/testing/_internal/common_quantization.py [hunk 3]
```diff
@@ -358,6 +367,65 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
 
         return models[False]
 
+    def checkGraphModuleHasNode(self, graph_module, target_node):
+        """ Check if GraphModule contains the target node
+        Args:
+            graph_module: the GraphModule instance we want to check
+            target_op: a tuple of 2 elements, first element is
```
Comment (supriyar) [inline torch/testing/_internal/common_quantization.py:374]: s/target_op/target_node

File: test/quantization/test_quantize_fx.py [hunk 4]
```diff
@@ -102,15 +112,74 @@ def forward(self, x):
...
+                else:
+                    self.relu = torch.nn.Identity()
+
+            def forward(self, x):
+                return self.relu(F.linear(x, self.w, self.b))
+
+        data = [torch.rand((1, 30), dtype=torch.float)]
+        options = itertools.product(
+            [(ModuleLinear(has_relu=False), True),
+             (FuncLinear(has_relu=False), False)],
```
Comment (supriyar) [inline test/quantization/test_quantize_fx.py:160]: are we missing a check for `has_relu=True`?

**Review by supriyar — APPROVED** [review]
lgtm, left a few comments

File: test/quantization/test_quantize_fx.py (hunk 4 above)
Comment (jerryzh168) [reply test/quantization/test_quantize_fx.py:160]: we do have it in L181, although some of the functionalities(functional linear + relu) are not supported yet

Comment (supriyar) [reply test/quantization/test_quantize_fx.py:160]: I see, I think we also need to add linear dynamic + relu fusion as well.

File: torch/quantization/fx/quantize.py [hunk 5]
```diff
@@ -565,6 +566,8 @@ def prepare_dynamic(self, model, qconfig_dict, inplace=False):
 
     def convert(self, observed, inplace=False, debug=False):
         assert self.activation_post_process_map is not None
+        # move to cpu since we only have quantized cpu kernels
+        observed.eval().cpu()
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:570]: thoughts on if we can let the user do this instead, and assert that the model is on CPU (or handle GPU)?  I think doing it here might be surprising to users

File: torch/testing/_internal/common_quantization.py [hunk 6]
```diff
@@ -358,6 +367,82 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
 
         return models[False]
 
+    def checkGraphModuleHasNode(self, graph_module, target_node):
+        """ Check if GraphModule contains the target node
+        Args:
+            graph_module: the GraphModule instance we want to check
+            target_node: a tuple of 2 elements, first element is
```
Comment (vkuzo) [inline torch/testing/_internal/common_quantization.py:374]: thanks for the docblock! can we provide an example call as well?

File: torch/testing/_internal/common_quantization.py [hunk 7]
```diff
@@ -358,6 +367,82 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
...
+        modules = dict(graph_module.root.named_modules())
+        for node in graph_module.graph.nodes:
+            if node.op == 'call_function' and node.op == target_node[0] and node.target == target_node[1]:
+                return
+            elif node.op == 'call_module' and node.op == target_node[0] and type(modules[node.target]) == target_node[1]:
+                return
+        self.assertTrue(False, 'node:' + str(target_node) +
+                        ' not found in the graph module')
+
+    def printGraphModule(self, graph_module):
```
Comment (vkuzo) [inline torch/testing/_internal/common_quantization.py:391]: not feedback for this PR, more just wondering if we can move this to GraphModule core at some point?

File: torch/testing/_internal/common_quantization.py [hunk 8]
```diff
@@ -358,6 +367,82 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
...
+        modules = dict(graph_module.root.named_modules())
+        for n in graph_module.graph.nodes:
+            node_info = ' '.join(map(repr, [n.op, n.name, n.target, n.args, n.kwargs]))
+            if n.op == 'call_module':
+                node_info += ' module type:' + repr(type(modules[n.target]))
+            print(node_info)
+
+    def checkGraphModeFxOp(self, model, inputs, quantized_node, quant_type=QuantType.STATIC, debug=False):
```
Comment (vkuzo) [inline torch/testing/_internal/common_quantization.py:399]: optional: maybe the name could be `quantWithFxAndCheckQuantNode` or something like that?

File: torch/testing/_internal/common_quantization.py [hunk 9]
```diff
@@ -358,6 +367,82 @@ def checkGraphModeOp(self, module, inputs, quantized_op, tracing=False, debug=Fa
...
+        fused = fuse(original)
+
+        quantizer = Quantizer()
+        # TODO: uncommon after we make per channel observer work in the flow
+        # qconfig_dict = {'': get_default_qconfig(torch.backends.quantized.engine)}
+        qconfig_dict = {'' : default_qconfig}
+        if quant_type == QuantType.DYNAMIC:
+            prepared = quantizer.prepare_dynamic(fused, qconfig_dict)
+        else:
+            prepared = quantizer.prepare(fused, qconfig_dict)
```
Comment (vkuzo) [inline torch/testing/_internal/common_quantization.py:425]: not feedback for this PR, more a question, should we have `prepare_qat` to be consistent with eager mode?

**Review by vkuzo — APPROVED** [review]
